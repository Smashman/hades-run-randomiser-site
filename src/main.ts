import { Run, RunOptions } from './run';

const runOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        heatRange: {
            min: 4,
            max: 12,
        },
    }
};

const run = Run.generateRandomRun(runOptions);

console.log('run', run);
console.log('run.weapon', run.weapon);
console.log('run.weaponAspect', run.weaponAspect);
console.log('run.keepsake', run.keepsake);
console.log('run.companion', run.companion);
console.log('run.mirror', run.mirror);
console.log('run.pact', run.pact);