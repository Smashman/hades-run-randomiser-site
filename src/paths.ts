export const imgPath = 'public/img';
export const unknownIconPath = `${imgPath}/misc/unknown_icon.png`;