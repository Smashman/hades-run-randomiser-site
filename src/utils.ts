export const getRandomItemFromArray = <T>(array: T[]) => array[Math.floor(Math.random() * array.length)];
export const fillArray = <T>(length: number, fillValue: T) => Array.from<T>({length: length}).fill(fillValue);

export interface Options {
    enabled?: boolean;
}

export abstract class Item {
    isUnlocked: boolean = false;

    unlock() {
        this.isUnlocked = true;
    }

    lock() {
        this.isUnlocked = false;
    }
}

export abstract class ItemList<T extends Item> {
    items: T[] = [];

    addItem(item: T) {
        this.items = [...this.items, item];
        return this;
    }

    addItems(items: T[]) {
        items.forEach(item => this.addItem(item));
        return this;
    }

    get unlocked() {
        return this.items.filter(item => item.isUnlocked);
    }

    unlockAll() {
        this.items.forEach(item => item.unlock());
    }

    getRandom(): T {
        return getRandomItemFromArray(this.unlocked) ?? null;
    }
}

export class Level {
    constructor(protected _value: number, readonly min: number, readonly max: number) {}

    set value(setTo: number) {
        if (setTo >= this.max) {
            this._value = this.max;
        }
        else if (setTo <= this.min) {
            this._value = this.min;
        }
        else {
            this._value = setTo;
        }
    }

    get value() {
        return this._value;
    }

    isMinLevel(): boolean {
        return this.value <= this.min;
    }

    isMaxLevel(): boolean {
        return this.value >= this.max;
    }

    setMaxLevel(): void {
        this.value = this.max;
    }

    addValue(change: number): number {
        this.value = this.value + change;
        return this.value;
    }

    increment(): number {
        return this.addValue(1);
    }

    decrement(): number {
        return this.addValue(-1);
    }
}