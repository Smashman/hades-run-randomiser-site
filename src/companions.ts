import { getRandomItemFromArray, Item, Options } from './utils';
import { ItemList } from './utils';

export class Companion extends Item {
    level: number = 1;
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level === Companion.maxLevel;
    }
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level = Companion.maxLevel);
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}