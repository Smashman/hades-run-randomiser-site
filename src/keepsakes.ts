import { getRandomItemFromArray, Options, ItemList, Item, Level } from './utils';

export interface KeepsakeOptions extends Options {
    ignoreMaxLevel?: boolean;
    changeBetweenBiomes?: boolean;
}

export class Keepsakes extends ItemList<Keepsake> {
    getRandomKeepsakes(keepsakeOptions?: KeepsakeOptions): KeepsakeConfiguration {
        const keepsakes = (keepsakeOptions?.ignoreMaxLevel && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(keepsake => !keepsake.isMaxLevel)] : [...this.unlocked];

        if (keepsakes.length === 0) {
            return null;
        }

        let keepsakesChosen: Keepsake[] = [getRandomItemFromArray(keepsakes)];

        if (keepsakeOptions?.changeBetweenBiomes) {
            for (let i = 1; i <= 3; i++) {
                keepsakesChosen = [...keepsakesChosen, getRandomItemFromArray(keepsakes.filter(keepsake => !keepsakesChosen.includes(keepsake))) ?? null];
            }
        }

        return keepsakesChosen as KeepsakeConfiguration;
    }

    maxLevelAll() {
        this.items.forEach(keepsake => keepsake.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(keepsake => keepsake.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(keepsake => keepsake.isMaxLevel);
    }
}

export class Keepsake extends Item {
    level: Level = new Level(1, Keepsake.minLevel, Keepsake.maxLevel);
    constructor(public name: string, public giver: string, public icon: string, public isHidden: boolean = false) {
        super();
    }

    static minLevel = 1;
    static maxLevel = 3;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }
}

export type KeepsakeConfiguration = [Keepsake] | [Keepsake, Keepsake, Keepsake, Keepsake] | null;