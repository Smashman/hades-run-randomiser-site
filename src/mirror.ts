import { getRandomItemFromArray, Item, Options, StorableItemData } from './utils';

export interface MirrorOptions extends Options {
};

export const defaultMirrorOptions: MirrorOptions = {
    enabled: true,
};

export class Mirror {
    groups: MirrorGroup[] = [];

    addGroup(sets: [[MirrorTalent, MirrorTalent], [MirrorTalent, MirrorTalent]], isUnlocked: boolean = false) {
        const mirrorTalentSets = <[MirrorTalentSet, MirrorTalentSet]>sets.map(set => new MirrorTalentSet(set));
        this.groups = [...this.groups, new MirrorGroup(mirrorTalentSets, isUnlocked)];
        return this;
    }

    get unlockedGroups() {
        return this.groups.filter(group => group.isUnlocked);
    }

    unlockAll() {
        this.groups.forEach(group => group.unlock());
    }

    getRandom(): MirrorConfiguration {
        return new MirrorConfiguration(this.unlockedGroups.map(group => group.getRandom()).reduce((acc, val) => [...acc, ...val], <MirrorTalent[]>[]));
    }

    toStorableData() {
        return {
            groups: this.groups.map(group => group.toStorableData())
        }
    }
}

export class MirrorGroup extends Item{
    constructor(public sets: [MirrorTalentSet, MirrorTalentSet], public isUnlocked: boolean = false) {
        super();
    }

    getRandom() {
        return <[MirrorTalent, MirrorTalent]>this.sets.map(set => set.getRandom());
    }
}

export class MirrorTalentSet {
    constructor(public talents: [MirrorTalent, MirrorTalent]) {
        talents[0].colour = "red";
        talents[1].colour = "green";
    }

    getRandom(): MirrorTalent {
        return getRandomItemFromArray(this.talents);
    }
}

export class MirrorTalent {
    rankCount: number;
    colour: "red" | "green";
    totalCost: number;

    constructor(public name: string, public rankCosts: number[]) {
        this.rankCount = rankCosts.length;
        this.totalCost = rankCosts.reduce((totalCost, cost) => totalCost + cost);
    }
}

export class MirrorConfiguration {
    totalDarkness: number;
    constructor(public talents: MirrorTalent[]){
        this.totalDarkness = talents.reduce((totalCost, talent) => totalCost + talent.totalCost, 0);
    };
}