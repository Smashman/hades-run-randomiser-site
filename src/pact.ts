import { getRandomItemFromArray, Options } from './utils';

export interface HeatRange {
    min?: number;
    max?: number;
}

export interface PactOptions extends Options {
    heatRange?: HeatRange
}

export class Pact {
    conditions: PactCondition[] = [];

    addCondition(pactCondition: PactCondition) {
        this.conditions = [...this.conditions, pactCondition];
        return this;
    }

    getRandomWithRange(heatRange?: HeatRange) {
        return PactConfiguration.generateFromHeatRange(this, heatRange);
    }

    unlockAll() {
        this.conditions.forEach(condition => condition.enforceLimit = false);
    }

    get maxHeat(): number {
        return this.conditions.reduce((maxTotal, condition) => maxTotal + condition.totalHeat, 0);
    }

    get initialHeat(): number {
        return this.conditions.reduce((total, condition) => total + condition.initialHeat, 0);
    }
}

export class PactCondition {
    enforceLimit: boolean = false;
    private _initialRank: number = 0;
    constructor(public name: string, public readonly rankCosts: number[], public rankLimit?: number) {
        this.enforceLimit = !!rankLimit;
    }

    get initialRank(): number {
        return this._initialRank;
    }

    set initialRank(rank: number) {
        this._initialRank = rank > this.rankCount ? this.rankCount : rank;
    }

    get rankCount(): number {
        return this.rankLimit && this.enforceLimit ? this.rankLimit : this.rankCosts.length;
    }

    get totalHeat(): number {
        return (this.rankLimit && this.enforceLimit ? this.rankCosts.slice(0, this.rankLimit) : this.rankCosts).reduce((totalHeat, cost) => totalHeat + cost);
    }

    get initialHeat(): number {
        return this.rankCosts.slice(0, this.initialRank).reduce((total, value) => total + value, 0);
    }
}

export class PactConfiguration {
    conditionConfigurations: PactConditionConfiguration[];
    heatTotal: number = 0;
    heatTarget: number = 0;
    heatRange: HeatRange;

    constructor(pact: Pact, heatRange?: HeatRange, ranks?: number[]) {
        this.conditionConfigurations = pact.conditions.map((condition, index) => new PactConditionConfiguration(condition, (ranks && ranks[index]) || null));
        this.heatRange = {
            min: heatRange?.min || pact.initialHeat,
            max: heatRange?.max || pact.maxHeat,
        };
    }

    static generateFromHeatRange(pact: Pact, heatRange?: HeatRange) {
        const pactConfiguration = new PactConfiguration(pact, heatRange);

        const configHeatRange = pactConfiguration.heatRange;


        const target = Math.floor(Math.random() * (configHeatRange.max - configHeatRange.min + 1) + configHeatRange.min);

        let current = pact.initialHeat;

        let remainingConditions = [...pactConfiguration.conditionConfigurations.filter((conditionConfig) => conditionConfig.rank !== conditionConfig.condition.rankCount)];


        while (current < target) {
            const randomCondition = getRandomItemFromArray(remainingConditions);
            current += randomCondition.rankUp();

            if (randomCondition.rank === randomCondition.condition.rankCount) {
                remainingConditions = [...remainingConditions.filter(condition => randomCondition !== condition)];
            }
        }

        pactConfiguration.heatTotal = current;
        pactConfiguration.heatTarget = target;

        return pactConfiguration;
    }
}

export class PactConditionConfiguration {
    private _rank: number = 0;

    constructor(public condition: PactCondition, rank?: number) {
        this.rank = rank || condition.initialRank;
    }

    get heat(): number {
        return this.condition.rankCosts.slice(0, this.rank).reduce((total, value) => total + value, 0);
    }

    get rank(): number {
        return this._rank;
    }

    set rank(rank: number) {
        this._rank = rank > this.condition.rankCount ? this.condition.rankCount : rank;
    }

    rankUp(): number {
        if (this.rank < this.condition.rankCount) {
            this.rank++;
            return this.condition.rankCosts[this.rank-1];
        }
        else {
            return null;
        }
    }
}