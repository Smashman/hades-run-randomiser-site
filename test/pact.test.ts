import { Pact, PactCondition, PactConditionConfiguration, PactConfiguration, HeatRange } from '../src/pact';
import * as utils from '../src/utils';

describe('PactCondition class', () => {
    it('should create expected PactCondition class instance', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);

        expect(pactCondition).toMatchSnapshot();
        expect(pactCondition.initialRank).toBe(0);
        expect(pactCondition.rankCount).toBe(5);
        expect(pactCondition.totalHeat).toBe(8);
        expect(pactCondition.initialHeat).toBe(0);
    });

    it('should create expected PactCondition class instance with rankLimit', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3], 4);

        expect(pactCondition).toMatchSnapshot();
        expect(pactCondition.initialRank).toBe(0);
        expect(pactCondition.rankCount).toBe(4);
        expect(pactCondition.totalHeat).toBe(5);
        expect(pactCondition.initialHeat).toBe(0);
    });

    it('should return expected initialHeat when initialRank is defined', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        pactCondition.initialRank = 2;

        expect(pactCondition.initialRank).toBe(2);
        expect(pactCondition.initialHeat).toBe(2);
    });

    it('should set initialRank to highest possible rank if number is higher than number of ranks', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        pactCondition.initialRank = 50;

        expect(pactCondition.initialRank).toBe(5);
        expect(pactCondition.initialHeat).toBe(8);
    });
});

describe('PactConditionConfiguration class', () => {
    it('should create expected PactConditionConfiguration class instance', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        expect(pactConditionConfiguration).toMatchSnapshot();
        expect(pactConditionConfiguration.rank).toBe(0);
        expect(pactConditionConfiguration.heat).toBe(0);
    });

    it('should create expected PactConditionConfiguration class instance with rank parameter', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition, 3);

        expect(pactConditionConfiguration).toMatchSnapshot();
        expect(pactConditionConfiguration.rank).toBe(3);
        expect(pactConditionConfiguration.heat).toBe(3);
    });

    it('should create expected PactConditionConfiguration class instance when pact condition has an initialRank', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        pactCondition.initialRank = 4;
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        expect(pactConditionConfiguration).toMatchSnapshot();
        expect(pactConditionConfiguration.rank).toBe(4);
        expect(pactConditionConfiguration.heat).toBe(5);
    });

    it('should create expected PactConditionConfiguration class instance when pact condition has an initialRank and rank parameter is set', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        pactCondition.initialRank = 4;
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition, 2);

        expect(pactConditionConfiguration).toMatchSnapshot();
        expect(pactConditionConfiguration.rank).toBe(2);
        expect(pactConditionConfiguration.heat).toBe(2);
    });

    it('should set rank', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        pactConditionConfiguration.rank = 2;

        expect(pactConditionConfiguration.rank).toBe(2);
        expect(pactConditionConfiguration.heat).toBe(2);
    });

    it('should set rank to maxiumum value when setting to number higher than maximum', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        pactConditionConfiguration.rank = 12;

        expect(pactConditionConfiguration.rank).toBe(5);
    });

    it('should increase heat and rank when rankUp is called', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        const rankUp1 = pactConditionConfiguration.rankUp();

        expect(rankUp1).toBe(1);
        expect(pactConditionConfiguration.rank).toBe(1);
        expect(pactConditionConfiguration.heat).toBe(1);

        const rankUp2 = pactConditionConfiguration.rankUp();

        expect(rankUp2).toBe(1);
        expect(pactConditionConfiguration.rank).toBe(2);
        expect(pactConditionConfiguration.heat).toBe(2);

        const rankUp3 = pactConditionConfiguration.rankUp();

        expect(rankUp3).toBe(1);
        expect(pactConditionConfiguration.rank).toBe(3);
        expect(pactConditionConfiguration.heat).toBe(3);

        const rankUp4 = pactConditionConfiguration.rankUp();

        expect(rankUp4).toBe(2);
        expect(pactConditionConfiguration.rank).toBe(4);
        expect(pactConditionConfiguration.heat).toBe(5);

        const rankUp5 = pactConditionConfiguration.rankUp();

        expect(rankUp5).toBe(3);
        expect(pactConditionConfiguration.rank).toBe(5);
        expect(pactConditionConfiguration.heat).toBe(8);
    });

    it('should return null and not increment rank at maxiumum rank', () => {
        const pactCondition = new PactCondition('Extra Angry', [1, 1, 1, 2, 3]);
        const pactConditionConfiguration = new PactConditionConfiguration(pactCondition);

        pactConditionConfiguration.rankUp();
        pactConditionConfiguration.rankUp();
        pactConditionConfiguration.rankUp();
        pactConditionConfiguration.rankUp();
        pactConditionConfiguration.rankUp();

        expect(pactConditionConfiguration.rank).toBe(5);
        expect(pactConditionConfiguration.heat).toBe(8);

        const rankUp6 = pactConditionConfiguration.rankUp();

        expect(rankUp6).toBe(null);
        expect(pactConditionConfiguration.rank).toBe(5);
        expect(pactConditionConfiguration.heat).toBe(8);
    });
});

describe('Pact class', () => {
    it('should create expected Pact class instance', () => {
        const pact = new Pact();

        expect(pact).toMatchSnapshot();
        expect(pact.maxHeat).toBe(0);
        expect(pact.initialHeat).toBe(0);
    });

    it('should add PactCondition to Pact', () => {
        const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3]));

        expect(pact).toMatchSnapshot();
        expect(pact.maxHeat).toBe(8);
        expect(pact.initialHeat).toBe(0);
    });

    it('should disable enforce limit on all conditions with unlockAll', () => {
        const pactCondition1 = new PactCondition('Extra Angry', [1, 1, 1, 2, 3], 4);
        const pactCondition2 = new PactCondition('Bigger Slam', [2, 2], 1);
        const pact = new Pact().addCondition(pactCondition1).addCondition(pactCondition2);

        expect(pactCondition1.enforceLimit).toBe(true);
        expect(pactCondition2.enforceLimit).toBe(true);
        expect(pact.maxHeat).toBe(7);

        pact.unlockAll();

        expect(pactCondition1.enforceLimit).toBe(false);
        expect(pactCondition2.enforceLimit).toBe(false);
        expect(pact.maxHeat).toBe(12);
    });

    it('should call PactConfiguration.generateFromHeatRange when calling getRandomWithRange', () => {
        const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
        const heatRange: HeatRange = {
            min: 1,
            max: 3,
        };

        const pactConfig = new PactConfiguration(pact, heatRange);
        const generateFromHeatRangeSpy = jest.spyOn(PactConfiguration, 'generateFromHeatRange').mockReturnValue(pactConfig);

        pact.getRandomWithRange(heatRange);

        expect(generateFromHeatRangeSpy).toBeCalledWith(pact, heatRange);

        generateFromHeatRangeSpy.mockRestore();
    });
});

describe('PactConfiguration class', () => {
    it('should create expected PactConfiguration class instance', () => {
        const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
        const pactConfig = new PactConfiguration(pact);

        expect(pactConfig).toMatchSnapshot();
    });

    it('should create expected PactConfiguration class instance with heatRange', () => {
        const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
        const heatRange: HeatRange = {
            min: 1,
            max: 3,
        };
        const pactConfig = new PactConfiguration(pact, heatRange);

        expect(pactConfig).toMatchSnapshot();
    });

    it('should create expected PactConfiguration class instance with ranks', () => {
        const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
        const ranks: number[] = [1, 1];
        const pactConfig = new PactConfiguration(pact, null, ranks);

        expect(pactConfig).toMatchSnapshot();
    });

    describe('generateFromHeatRange', () => {
        let randomMock: jest.SpyInstance;
        let index: number;
        let getRandomItemFromArrayMock: jest.SpyInstance;
        const randomIndexArray = [1, 0, 2, 0];

        beforeEach(() => {
            randomMock = jest.spyOn(global.Math, 'random').mockReturnValue(0.5);

            index = 0;
            getRandomItemFromArrayMock = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(array => {
                if (index >= randomIndexArray.length) {
                    index = 0;
                }
                return array[randomIndexArray[index++]];
            });
        });

        afterEach(() => {
            getRandomItemFromArrayMock.mockRestore();
            randomMock.mockRestore();
        });

        it('should return expected PactConfiguration class instance', () => {
            const pact = new Pact()
                .addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3]))
                .addCondition(new PactCondition('Bigger Slam', [2, 2]))
                .addCondition(new PactCondition('Harder Skin', [5]));
            const randomPactConfiguration = pact.getRandomWithRange();

            const expectedHeatRange: HeatRange = {
                min: 0,
                max: 17,
            };

            expect(randomPactConfiguration).toMatchSnapshot();
            expect(randomPactConfiguration.heatRange).toStrictEqual(expectedHeatRange);
            expect(randomPactConfiguration.heatTarget).toBe(9);
            expect(randomPactConfiguration.heatTotal).toBe(9);

            expect(randomMock).toHaveBeenCalledTimes(1);
            expect(getRandomItemFromArrayMock).toHaveBeenCalledTimes(4);
        });

        it('should return expected PactConfiguration class instance with heatRange', () => {
            const pact = new Pact()
                .addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3]))
                .addCondition(new PactCondition('Bigger Slam', [2, 2]))
                .addCondition(new PactCondition('Harder Skin', [5]));
            const heatRange: HeatRange = {
                min: 1,
                max: 3,
            };
            const randomPactConfiguration = pact.getRandomWithRange(heatRange);

            expect(randomPactConfiguration).toMatchSnapshot();
            expect(randomPactConfiguration.heatRange).toStrictEqual(heatRange);
            expect(randomPactConfiguration.heatTarget).toBe(2);
            expect(randomPactConfiguration.heatTotal).toBe(2);

            expect(randomMock).toHaveBeenCalledTimes(1);
            expect(getRandomItemFromArrayMock).toHaveBeenCalledTimes(1);
        });
    });
});