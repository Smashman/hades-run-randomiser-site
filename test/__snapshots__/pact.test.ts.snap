// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pact class should add PactCondition to Pact 1`] = `
Pact {
  "conditions": Array [
    PactCondition {
      "_initialRank": 0,
      "enforceLimit": false,
      "name": "Extra Angry",
      "rankCosts": Array [
        1,
        1,
        1,
        2,
        3,
      ],
      "rankLimit": undefined,
    },
  ],
}
`;

exports[`Pact class should create expected Pact class instance 1`] = `
Pact {
  "conditions": Array [],
}
`;

exports[`PactCondition class should create expected PactCondition class instance 1`] = `
PactCondition {
  "_initialRank": 0,
  "enforceLimit": false,
  "name": "Extra Angry",
  "rankCosts": Array [
    1,
    1,
    1,
    2,
    3,
  ],
  "rankLimit": undefined,
}
`;

exports[`PactCondition class should create expected PactCondition class instance with rankLimit 1`] = `
PactCondition {
  "_initialRank": 0,
  "enforceLimit": true,
  "name": "Extra Angry",
  "rankCosts": Array [
    1,
    1,
    1,
    2,
    3,
  ],
  "rankLimit": 4,
}
`;

exports[`PactConditionConfiguration class should create expected PactConditionConfiguration class instance 1`] = `
PactConditionConfiguration {
  "_rank": 0,
  "condition": PactCondition {
    "_initialRank": 0,
    "enforceLimit": false,
    "name": "Extra Angry",
    "rankCosts": Array [
      1,
      1,
      1,
      2,
      3,
    ],
    "rankLimit": undefined,
  },
}
`;

exports[`PactConditionConfiguration class should create expected PactConditionConfiguration class instance when pact condition has an initialRank 1`] = `
PactConditionConfiguration {
  "_rank": 4,
  "condition": PactCondition {
    "_initialRank": 4,
    "enforceLimit": false,
    "name": "Extra Angry",
    "rankCosts": Array [
      1,
      1,
      1,
      2,
      3,
    ],
    "rankLimit": undefined,
  },
}
`;

exports[`PactConditionConfiguration class should create expected PactConditionConfiguration class instance when pact condition has an initialRank and rank parameter is set 1`] = `
PactConditionConfiguration {
  "_rank": 2,
  "condition": PactCondition {
    "_initialRank": 4,
    "enforceLimit": false,
    "name": "Extra Angry",
    "rankCosts": Array [
      1,
      1,
      1,
      2,
      3,
    ],
    "rankLimit": undefined,
  },
}
`;

exports[`PactConditionConfiguration class should create expected PactConditionConfiguration class instance with rank parameter 1`] = `
PactConditionConfiguration {
  "_rank": 3,
  "condition": PactCondition {
    "_initialRank": 0,
    "enforceLimit": false,
    "name": "Extra Angry",
    "rankCosts": Array [
      1,
      1,
      1,
      2,
      3,
    ],
    "rankLimit": undefined,
  },
}
`;

exports[`PactConfiguration class generateFromHeatRange should return expected PactConfiguration class instance 1`] = `
PactConfiguration {
  "conditionConfigurations": Array [
    PactConditionConfiguration {
      "_rank": 2,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Extra Angry",
        "rankCosts": Array [
          1,
          1,
          1,
          2,
          3,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 1,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Bigger Slam",
        "rankCosts": Array [
          2,
          2,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 1,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Harder Skin",
        "rankCosts": Array [
          5,
        ],
        "rankLimit": undefined,
      },
    },
  ],
  "heatRange": Object {
    "max": 17,
    "min": 0,
  },
  "heatTarget": 9,
  "heatTotal": 9,
}
`;

exports[`PactConfiguration class generateFromHeatRange should return expected PactConfiguration class instance with heatRange 1`] = `
PactConfiguration {
  "conditionConfigurations": Array [
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Extra Angry",
        "rankCosts": Array [
          1,
          1,
          1,
          2,
          3,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 1,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Bigger Slam",
        "rankCosts": Array [
          2,
          2,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Harder Skin",
        "rankCosts": Array [
          5,
        ],
        "rankLimit": undefined,
      },
    },
  ],
  "heatRange": Object {
    "max": 3,
    "min": 1,
  },
  "heatTarget": 2,
  "heatTotal": 2,
}
`;

exports[`PactConfiguration class should create expected PactConfiguration class instance 1`] = `
PactConfiguration {
  "conditionConfigurations": Array [
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Extra Angry",
        "rankCosts": Array [
          1,
          1,
          1,
          2,
          3,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Bigger Slam",
        "rankCosts": Array [
          2,
          2,
        ],
        "rankLimit": undefined,
      },
    },
  ],
  "heatRange": Object {
    "max": 12,
    "min": 0,
  },
  "heatTarget": 0,
  "heatTotal": 0,
}
`;

exports[`PactConfiguration class should create expected PactConfiguration class instance with heatRange 1`] = `
PactConfiguration {
  "conditionConfigurations": Array [
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Extra Angry",
        "rankCosts": Array [
          1,
          1,
          1,
          2,
          3,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 0,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Bigger Slam",
        "rankCosts": Array [
          2,
          2,
        ],
        "rankLimit": undefined,
      },
    },
  ],
  "heatRange": Object {
    "max": 3,
    "min": 1,
  },
  "heatTarget": 0,
  "heatTotal": 0,
}
`;

exports[`PactConfiguration class should create expected PactConfiguration class instance with ranks 1`] = `
PactConfiguration {
  "conditionConfigurations": Array [
    PactConditionConfiguration {
      "_rank": 1,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Extra Angry",
        "rankCosts": Array [
          1,
          1,
          1,
          2,
          3,
        ],
        "rankLimit": undefined,
      },
    },
    PactConditionConfiguration {
      "_rank": 1,
      "condition": PactCondition {
        "_initialRank": 0,
        "enforceLimit": false,
        "name": "Bigger Slam",
        "rankCosts": Array [
          2,
          2,
        ],
        "rankLimit": undefined,
      },
    },
  ],
  "heatRange": Object {
    "max": 12,
    "min": 0,
  },
  "heatTarget": 0,
  "heatTotal": 0,
}
`;
